Can't figure out `+` commutes in some contexts on GHC 8.6.3

The following compiles on GHC 8.4.4, but fails on 8.6.3:

```haskell
{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise #-}

module Test where

import Clash.Prelude

lastBits
  :: forall n a
   . ( BitPack a
     , n <= BitSize a
     , KnownNat (BitSize a)
     , KnownNat n
     )
  => a
  -> BitVector n
lastBits = leToPlus @n @(BitSize a) $ snd . split @_ @_ @n
```

where

```haskell
leToPlus
  :: forall (k :: Nat) (n :: Nat) r
   . ( k <= n
     )
  => (forall m . (n ~ (k + m)) => r)
  -- ^ Context with the (k + m ~ n) constraint
  -> r
leToPlus r = r @(n - k)
{-# INLINE leToPlus #-}
```

Changing `leToPlus` to:

```haskell
leToPlus
  :: forall (k :: Nat) (n :: Nat) r
   . ( k <= n
     )
  => (forall m . (n ~ (k + m), n ~ (m + k)) => r)
  -- ^ Context with the (k + m ~ n) constraint
  -> r
leToPlus r = r @(n - k)
{-# INLINE leToPlus #-}
```

..makes it compile just fine. The error reported by GHC is:

```
Test.hs:24:45: error:
    • Could not deduce: (w0 + n) ~ BitSize a
        arising from a use of ‘split’
      from the context: (BitPack a, n <= BitSize a, KnownNat (BitSize a),
                         KnownNat n)
        bound by the type signature for:
                   lastBits :: forall (n :: Nat) a.
                               (BitPack a, n <= BitSize a, KnownNat (BitSize a), KnownNat n) =>
                               a -> BitVector n
        at Test.hs:(15,1)-(23,16)
      or from: BitSize a ~ (n + m)
        bound by a type expected by the context:
                   forall (m :: Nat). (BitSize a ~ (n + m)) => a -> BitVector n
        at Test.hs:24:12-58
      The type variable ‘w0’ is ambiguous
    • In the second argument of ‘(.)’, namely ‘split @_ @_ @n’
      In the second argument of ‘($)’, namely ‘snd . split @_ @_ @n’
      In the expression: leToPlus @n @(BitSize a) $ snd . split @_ @_ @n
    • Relevant bindings include
        lastBits :: a -> BitVector n (bound at Test.hs:24:1)
   |
24 | lastBits = leToPlus @n @(BitSize a) $ snd . split @_ @_ @n
   |                                             ^^^^^^^^^^^^^^
```

For some reason, it sometimes can't figure out that `n + m` ~ `m + n`.

--- a/src/GHC/TypeLits/Normalise.hs
+++ b/src/GHC/TypeLits/Normalise.hs
@@ -195,9 +195,8 @@ import TcPluginM  (TcPluginM, tcPluginTrace, tcPluginIO)
 import Type       (Kind, PredType, eqType, mkTyVarTy)
 import TysWiredIn (typeNatKind)
 
-import Coercion   (CoercionHole, Role (..), mkForAllCos, mkHoleCo, mkInstCo,
-                   mkNomReflCo, mkUnivCo)
-import TcPluginM  (newCoercionHole, newFlexiTyVar, tcLookupClass)
+import Coercion   (CoercionHole, Role (..), mkUnivCo)
+import TcPluginM  (newCoercionHole, tcLookupClass)
 import TcRnTypes  (TcPlugin (..), TcPluginResult(..))
 import TyCoRep    (UnivCoProvenance (..))
 import TcType     (isEqPred)
@@ -373,7 +372,7 @@ decideEqualSOP opts gen'd givens  _deriveds wanteds = do
         sr <- simplifyNats opts unit_givens unit_wanteds
         tcPluginTrace "normalised" (ppr sr)
         reds <- forM reducible_wanteds $ \(origCt,(term, ws)) -> do
-          wants <- fmap fst $ evSubtPreds origCt $ subToPred opts ws
+          wants <- evSubtPreds origCt $ subToPred opts ws
           return ((term, origCt), wants)
         case sr of
           Simplified evs -> do
@@ -658,7 +657,7 @@ unifyItemToPredType ui =
             SubstItem {..} -> reifySOP siSOP
             UnifyItem {..} -> reifySOP siRHS
 
-evSubtPreds :: Ct -> [(PredType,Kind)] -> TcPluginM ([Ct], [CoercionHole])
+evSubtPreds :: Ct -> [(PredType,Kind)] -> TcPluginM [Ct]
 evSubtPreds ct preds = do
   let predTypes = map fst preds
 #if MIN_VERSION_ghc(8,4,1)
@@ -666,34 +665,21 @@ evSubtPreds ct preds = do
 #else
   holes <- replicateM (length preds) newCoercionHole
 #endif
-  return (zipWith (unifyItemToCt (ctLoc ct)) predTypes holes, holes)
+  return (zipWith (unifyItemToCt (ctLoc ct)) predTypes holes)
 
 evMagic :: Ct -> Set CType -> [(PredType,Kind)] -> TcPluginM (Maybe ((EvTerm, Ct), [Ct]))
 evMagic ct knW preds = case classifyPredType $ ctEvPred $ ctEvidence ct of
   EqPred NomEq t1 t2 -> do
-    (holeWanteds, holes) <- evSubtPreds ct preds
-    let predKinds = map snd preds
+    holeWanteds <- evSubtPreds ct preds
     knWanted <- mapM (mkKnWanted ct) (toList knW)
     let newWant = knWanted ++ holeWanteds
-        ctEv      = mkUnivCo (PluginProv "ghc-typelits-natnormalise") Nominal t1 t2
-#if MIN_VERSION_ghc(8,4,1)
-        holeEvs   = map mkHoleCo holes
-#else
-        predTypes = map fst preds
-        holeEvs   = zipWith (\h p -> uncurry (mkHoleCo h Nominal) (getEqPredTys p)) holes predTypes
-#endif
-    forallEv <- mkForAllCos <$> mapM mkCoVar predKinds <*> pure ctEv
-    let finalEv = foldl mkInstCo forallEv holeEvs
+        ctEv    = mkUnivCo (PluginProv "ghc-typelits-natnormalise") Nominal t1 t2
 #if MIN_VERSION_ghc(8,5,0)
-    return (Just ((EvExpr (Coercion finalEv), ct),newWant))
+    return (Just ((EvExpr (Coercion ctEv), ct),newWant))
 #else
-    return (Just ((EvCoercion finalEv, ct),newWant))
+    return (Just ((EvCoercion ctEv, ct),newWant))
 #endif
   _ -> return Nothing
-  where
-    mkCoVar k = (,natReflCo) <$> newFlexiTyVar k
-      where
-        natReflCo = mkNomReflCo k
 
 mkNonCanonical' :: CtLoc -> CtEvidence -> Ct
 mkNonCanonical' origCtl ev =

--- a/src/GHC/TypeLits/Normalise.hs
+++ b/src/GHC/TypeLits/Normalise.hs
@@ -163,8 +163,8 @@ import Control.Monad       ((<=<), forM)
 import Control.Monad       (replicateM)
 #endif
 import Control.Monad.Trans.Writer.Strict
-import Data.Either         (rights)
-import Data.List           (intersect, stripPrefix, find)
+import Data.Either         (partitionEithers, rights)
+import Data.List           (intersect, partition, stripPrefix, find)
 import Data.Maybe          (mapMaybe, catMaybes)
 import Data.Set            (Set, empty, toList, notMember, fromList, union)
 import GHC.TcPluginM.Extra (tracePlugin, newGiven, newWanted)
@@ -212,7 +212,8 @@ import Data.IORef
 #if MIN_VERSION_ghc(8,10,0)
 import Constraint
   (Ct, CtEvidence (..), CtLoc, TcEvDest (..), ctEvidence, ctEvLoc, ctEvPred,
-   ctLoc, ctLocSpan, isGiven, isWanted, mkNonCanonical, setCtLoc, setCtLocSpan)
+   ctLoc, ctLocSpan, isGiven, isWanted, mkNonCanonical, setCtLoc, setCtLocSpan,
+   isWantedCt)
 import Predicate
   (EqRel (NomEq), Pred (EqPred), classifyPredType, getEqPredTys, mkClassPred,
    mkPrimEqPred)
@@ -220,7 +221,8 @@ import Type (typeKind)
 #else
 import TcRnTypes
   (Ct, CtEvidence (..), CtLoc, TcEvDest (..), ctEvidence, ctEvLoc, ctEvPred,
-   ctLoc, ctLocSpan, isGiven, isWanted, mkNonCanonical, setCtLoc, setCtLocSpan)
+   ctLoc, ctLocSpan, isGiven, isWanted, mkNonCanonical, setCtLoc, setCtLocSpan,
+   isWantedCt)
 import TcType (typeKind)
 import Type
   (EqRel (NomEq), PredTree (EqPred), classifyPredType, getEqPredTys, mkClassPred,
@@ -248,6 +250,7 @@ import TcType (isEqPrimPred)
 #endif
 
 -- internal
+import GHC.TypeLits.Normalise.SOP
 import GHC.TypeLits.Normalise.Unify
 
 #if !MIN_VERSION_ghc(8,10,0)
@@ -459,9 +462,23 @@ simplifyNats
   -> [(Either NatEquality NatInEquality,[(Type,Type)])]
   -- ^ Wanted constraints
   -> TcPluginM SimplifyResult
-simplifyNats opts@Opts {..} eqsG eqsW =
-    let eqs = map (second (const [])) eqsG ++ eqsW
-    in  tcPluginTrace "simplifyNats" (ppr eqs) >> simples [] [] [] [] eqs
+simplifyNats opts@Opts {..} eqsG eqsW = do
+    let eqsG1 = map (second (const ([] :: [(Type,Type)]))) eqsG
+        (varEqs,otherEqs) = partition isVarEqs eqsG1
+        fancyGivens = concatMap (makeGivensSet otherEqs) varEqs
+    case varEqs of
+      [] -> do
+        let eqs = otherEqs ++ eqsW
+        tcPluginTrace "simplifyNats" (ppr eqs)
+        simples [] [] [] [] eqs
+      _  -> do
+        tcPluginTrace ("simplifyNats(backtrack: " ++ show (length fancyGivens) ++ ")")
+                      (ppr varEqs)
+        foldr findFirstSimpliedWanted (Simplified []) <$>
+              mapM (\v -> do let eqs = v ++ eqsW
+                             tcPluginTrace "simplifyNats" (ppr eqs)
+                             simples [] [] [] [] eqs)
+              fancyGivens
   where
     simples :: [CoreUnify]
             -> [((EvTerm, Ct), [Ct])]
@@ -533,6 +550,51 @@ simplifyNats opts@Opts {..} eqsG eqsW =
 
     eqToLeq x y = [(x,y,True),(y,x,True)]
     substLeq s (x,y,b) = (substsSOP s x, substsSOP s y, b)
+
+    isVarEqs (Left (_,S [P [V _]], S [P [V _]]), _) = True
+    isVarEqs _ = False
+
+    makeGivensSet otherEqs varEq
+      = let (noMentionsV,mentionsV)   = partitionEithers
+                                          (map (matchesVarEq varEq) otherEqs)
+            (mentionsLHS,mentionsRHS) = partitionEithers mentionsV
+            vS = swapVar varEq
+            givensLHS = case mentionsLHS of
+              [] -> []
+              _  -> [mentionsLHS ++ ((varEq:mentionsRHS) ++ noMentionsV)]
+            givensRHS = case mentionsRHS of
+              [] -> []
+              _  -> [mentionsRHS ++ (vS:mentionsLHS ++ noMentionsV)]
+        in  case mentionsV of
+              [] -> [noMentionsV]
+              _  -> givensLHS ++ givensRHS
+
+    matchesVarEq (Left (_, S [P [V v1]], S [P [V v2]]),_) r = case r of
+      (Left (_,S [P [V v3]],_),_)
+        | v1 == v3 -> Right (Left r)
+        | v2 == v3 -> Right (Right r)
+      (Left (_,_,S [P [V v3]]),_)
+        | v1 == v3 -> Right (Left r)
+        | v2 == v3 -> Right (Right r)
+      (Right (_,(S [P [V v3]],_,_)),_)
+        | v1 == v3 -> Right (Left r)
+        | v2 == v3 -> Right (Right r)
+      (Right (_,(_,S [P [V v3]],_)),_)
+        | v1 == v3 -> Right (Left r)
+        | v2 == v3 -> Right (Right r)
+      _ -> Left r
+    matchesVarEq _ _ = error "internal error"
+
+    swapVar (Left (ct,S [P [V v1]], S [P [V v2]]),ps) =
+      (Left (ct,S [P [V v2]], S [P [V v1]]),ps)
+    swapVar _ = error "internal error"
+
+    findFirstSimpliedWanted (Impossible e)   _  = Impossible e
+    findFirstSimpliedWanted (Simplified evs) s2
+      | any (isWantedCt . snd . fst) evs
+      = Simplified evs
+      | otherwise
+      = s2
 
 -- If we allow negated numbers we simply do not emit the inequalities
 -- derived from the subtractions that are converted to additions with a

